File: sdk/packages/go/transport.go
package whodb

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"sync"
)

// Transport executes one platform operation and returns the GraphQL data
// object. Implementations: the default HTTP transport and IpcTransport
// (inside the WhoDB Functions runtime). The generated core and facades never
// speak HTTP directly.
type Transport interface {
	Execute(ctx context.Context, operation, document string, variables map[string]any) (map[string]any, error)
}

var retryableStatus = map[int]bool{502: true, 503: true, 504: true}

// httpTransport is the default GraphQL-over-HTTP transport (POST /api/query).
type httpTransport struct {
	endpoint    string
	credentials CredentialProvider
	client      *http.Client

	mu        sync.RWMutex
	orgID     string
	projectID string
}

func newHTTPTransport(host string, credentials CredentialProvider) *httpTransport {
	return &httpTransport{
		endpoint:    stripTrailingSlash(host) + "/api/query",
		credentials: credentials,
		client:      http.DefaultClient,
	}
}

func stripTrailingSlash(host string) string {
	for len(host) > 0 && host[len(host)-1] == '/' {
		host = host[:len(host)-1]
	}
	return host
}

// setWorkspace sets the workspace scope headers used on subsequent requests.
func (t *httpTransport) setWorkspace(orgID, projectID string) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.orgID = orgID
	t.projectID = projectID
}

func (t *httpTransport) post(ctx context.Context, body []byte) (*http.Response, error) {
	token, err := t.credentials.Token(ctx)
	if err != nil {
		return nil, err
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, t.endpoint, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	request.Header.Set("Content-Type", "application/json")
	request.Header.Set("Authorization", "Bearer "+token)
	request.Header.Set("User-Agent", "clidey-whodb-go/"+SDKVersion)
	t.mu.RLock()
	if t.orgID != "" {
		request.Header.Set("X-Whodb-Org-Id", t.orgID)
	}
	if t.projectID != "" {
		request.Header.Set("X-Whodb-Project-Id", t.projectID)
	}
	t.mu.RUnlock()
	return t.client.Do(request)
}

// Execute performs one operation with a single 401-refresh and 5xx retry.
func (t *httpTransport) Execute(ctx context.Context, operation, document string, variables map[string]any) (map[string]any, error) {
	body, err := json.Marshal(map[string]any{"query": document, "variables": variables})
	if err != nil {
		return nil, err
	}
	response, err := t.post(ctx, body)
	if err != nil {
		return nil, err
	}
	if response.StatusCode == http.StatusUnauthorized {
		_ = response.Body.Close()
		t.credentials.Refresh()
		if response, err = t.post(ctx, body); err != nil {
			return nil, err
		}
	} else if retryableStatus[response.StatusCode] {
		_ = response.Body.Close()
		if response, err = t.post(ctx, body); err != nil {
			return nil, err
		}
	}
	defer response.Body.Close()
	if response.StatusCode == http.StatusUnauthorized {
		return nil, fmt.Errorf("%w — check your API key or run: whodb login", ErrAuth)
	}
	if response.StatusCode >= 400 {
		return nil, &PlatformError{Message: fmt.Sprintf("platform request failed with HTTP %d", response.StatusCode), Code: fmt.Sprintf("HTTP_%d", response.StatusCode)}
	}
	var payload struct {
		Data   map[string]any `json:"data"`
		Errors []graphQLError `json:"errors"`
	}
	if err := json.NewDecoder(response.Body).Decode(&payload); err != nil {
		return nil, &PlatformError{Message: "invalid response for " + operation, Code: "INVALID_RESPONSE"}
	}
	if len(payload.Errors) > 0 {
		return nil, mapGraphQLErrors(payload.Errors)
	}
	if payload.Data == nil {
		return nil, &PlatformError{Message: "empty response for " + operation, Code: "EMPTY_RESPONSE"}
	}
	return payload.Data, nil
}

File: sdk/packages/rust/src/transport.rs
//! Transports: how prepared operations reach the platform. The generated
//! core and the facades never speak HTTP directly.

use serde_json::{json, Map, Value};
use std::sync::RwLock;

use crate::auth::CredentialProvider;
use crate::error::{map_graphql_errors, Error, Result};
use crate::SDK_VERSION;

/// Executes one platform operation, returning the GraphQL data object.
pub trait Transport: Send + Sync {
    /// Runs the operation and returns the `data` object keyed by operation name.
    fn execute(
        &self,
        operation: &str,
        document: &str,
        variables: Map<String, Value>,
    ) -> Result<Value>;
}

const RETRYABLE_STATUS: &[u16] = &[502, 503, 504];

/// Default GraphQL-over-HTTP transport (POST /api/query).
pub(crate) struct HttpTransport {
    endpoint: String,
    credentials: Box<dyn CredentialProvider>,
    workspace: RwLock<(String, String)>,
    agent: ureq::Agent,
}

impl HttpTransport {
    pub(crate) fn new(host: &str, credentials: Box<dyn CredentialProvider>) -> Self {
        HttpTransport {
            endpoint: format!("{}/api/query", host.trim_end_matches('/')),
            credentials,
            workspace: RwLock::new((String::new(), String::new())),
            agent: ureq::Agent::new_with_defaults(),
        }
    }

    pub(crate) fn set_workspace(&self, org_id: &str, project_id: &str) {
        let mut workspace = self.workspace.write().expect("workspace lock poisoned");
        *workspace = (org_id.to_string(), project_id.to_string());
    }

    fn post(&self, body: &Value) -> Result<(u16, Value)> {
        let token = self.credentials.token()?;
        let (org_id, project_id) = self
            .workspace
            .read()
            .expect("workspace lock poisoned")
            .clone();
        let mut request = self
            .agent
            .post(&self.endpoint)
            .header("Content-Type", "application/json")
            .header("Authorization", &format!("Bearer {token}"))
            .header("User-Agent", &format!("clidey-whodb-rust/{SDK_VERSION}"));
        if !org_id.is_empty() {
            request = request.header("X-Whodb-Org-Id", &org_id);
        }
        if !project_id.is_empty() {
            request = request.header("X-Whodb-Project-Id", &project_id);
        }
        match request.send_json(body) {
            Ok(mut response) => {
                let status = response.status().as_u16();
                let payload = response
                    .body_mut()
                    .read_json::<Value>()
                    .unwrap_or(Value::Null);
                Ok((status, payload))
            }
            Err(ureq::Error::StatusCode(code)) => Ok((code, Value::Null)),
            Err(error) => Err(Error::Transport(error.to_string())),
        }
    }
}

impl Transport for HttpTransport {
    fn execute(
        &self,
        operation: &str,
        document: &str,
        variables: Map<String, Value>,
    ) -> Result<Value> {
        let body = json!({"query": document, "variables": Value::Object(variables)});
        let (mut status, mut payload) = self.post(&body)?;
        if status == 401 {
            self.credentials.refresh();
            (status, payload) = self.post(&body)?;
        } else if RETRYABLE_STATUS.contains(&status) {
            (status, payload) = self.post(&body)?;
        }
        if status == 401 {
            return Err(Error::Auth(
                "check your API key or run: whodb login".to_string(),
            ));
        }
        if status >= 400 {
            return Err(Error::Platform {
                message: format!("platform request failed with HTTP {status}"),
                code: format!("HTTP_{status}"),
            });
        }
        if let Some(errors) = payload.get("errors").and_then(Value::as_array) {
            if !errors.is_empty() {
                return Err(map_graphql_errors(errors));
            }
        }
        match payload.get("data") {
            Some(data) if !data.is_null() => Ok(data.clone()),
            _ => Err(Error::Platform {
                message: format!("empty response for {operation}"),
                code: "EMPTY_RESPONSE".to_string(),
            }),
        }
    }
}
